gtda/diagrams/_metrics.py

```diff
@@ -74,55 +82,76 @@ def landscapes(diagrams, sampling, n_layers):
+        gaussian_filter(image, sigma_pixel, mode="constant", output=image)
+
+    heats_ -= np.transpose(heats_, (0, 2, 1))
+    heats_ /= (step_size ** 2)
     heats_ = np.rot90(heats_, k=1, axes=(1, 2))
     return heats_
 
 
-def persistence_images(diagrams, sampling, step_size, weights, sigma):
-    persistence_images_ = np.zeros(
-        (diagrams.shape[0], sampling.shape[0], sampling.shape[0]))
+def persistence_images(diagrams, sampling, step_size, sigma, weights):
+    # For persistence images, `sampling` is a tall matrix with two columns
+    # (the first for birth and the second for persistence), and `step_size` is
+    # a 2d array
+    # WARNING: modifies `diagrams` in place
+    persistence_images_ = \
+        np.zeros((len(diagrams), len(sampling), len(sampling)), dtype=float)
     # Transform diagrams from (birth, death, dim) to (birth, persistence, dim)
-    diagrams[:, :, 1] = diagrams[:, :, 1] - diagrams[:, :, 0]
-
-    for axis in [0, 1]:
-        # Set the values outside of the sampling range to the sampling range.
-        diagrams[:, :, axis][diagrams[:, :, axis] < sampling[0, axis]] = \
-            sampling[0, axis]
-        diagrams[:, :, axis][diagrams[:, :, axis] > sampling[-1, axis]] = \
-            sampling[-1, axis]
-        # Convert into pixel
-        diagrams[:, :, axis] = np.array(
-            (diagrams[:, :, axis] - sampling[0, axis]) / step_size[axis],
-            dtype=int)
-    # Sample the image
-    [_sample_image(persistence_images_[i], sampled_diag)
-     for i, sampled_diag in enumerate(diagrams)]
-
-    # Apply the weights
-    persistence_images_ *= weights / np.max(weights)
-
-    # Smoothen the weighted-image
-    for i, image in enumerate(persistence_images_):
-        persistence_images_[i] = gaussian_filter(image, sigma, mode="reflect")
+    diagrams[:, :, 1] -= diagrams[:, :, 0]
+
+    sigma_pixel = []
+    first_samplings = sampling[0]
+    last_samplings = sampling[-1]
+    for ax in [0, 1]:
+        diagrams_ax = diagrams[:, :, ax]
+        # Set the values outside of the sampling range
+        diagrams_ax[diagrams_ax < first_samplings[ax]] = first_samplings[ax]
+        diagrams_ax[diagrams_ax > last_samplings[ax]] = last_samplings[ax]
+        # Calculate the value of the component of `sigma` in pixel units
+        sigma_pixel.append(sigma / step_size[ax])
+
+    # diagrams_ax[...] = (diagrams_ax - first_sampling) / step_size[ax]
+    #
+    # diagrams = diagrams.astype(int)
+
+    # Sample the image, apply the weights, smoothen
+    for i, diagram in enumerate(diagrams):
+        nontrivial_points_idx = np.flatnonzero(diagram[:, 1])
+        diagram_nontrivial_pixel_coords = np.array(
+            (diagram - first_samplings) / step_size, dtype=int
+        )[nontrivial_points_idx]
+        image = persistence_images_[i]
+        _sample_image(image, diagram_nontrivial_pixel_coords)
+        image *= weights
+        gaussian_filter(image, sigma_pixel, mode="constant", output=image)
```

Comment:
Seems like the smoothing is applied twice? (lines 150-151 a well)

Reply:
Oh dear! Well spotted!!

Reply:
Should be fixed now.

Reply:
Should be fixed now.